cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,79 @@ cuML is thread safe so its functions can be called from multiple host threads if
+```cpp
+template <typename T>
+class DecisionTreeClassifier {
+  TreeNode<T>* root;
+  DTParams params;
+  const cumlHandle &handle;
+public:
+  DecisionTreeClassifier(const cumlHandle &handle, DTParams& params, bool verbose=false);
+  void fit(const T *input, int n_rows, int n_cols, const int *labels);
+  void predict(const T *input, int n_rows, int n_cols, int *predictions);
+};
+
+void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
+                               const int *labels, DecisionTreeClassifier<float> *model, DTParams params,
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const float* input,
+                                   DecisionTreeClassifier<float> *model, int n_rows,
+                                   int n_cols, int* predictions, bool verbose=false);
+```
+
+One of the right ways to expose this would be:
+```cpp
+// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
+// and passed between fit and predict methods
+template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
+struct DTParams { /* hyper-params for building DT */ };
+typedef TreeNode<float> TreeNodeF;
+typedef TreeNode<double> TreeNodeD;
+
+void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
+                               const int *labels, TreeNodeF *&root, DTParams params,
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const double* input, int n_rows,
+                                   int n_cols, const TreeNodeD *root, int* predictions,
+                                   bool verbose=false);
+```
+The above example obviously under-plays the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.
+
+### WIP TODO (Need to explore this further!)
+The above example also means, it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing this example, we could expose the following methods to achieve this:
+```cpp
+void storeTreeNode(const TreeNodeF *root, std::ostream &os);
+void storeTreeNode(const TreeNodeD *root, std::ostream &os);
+void loadTreeNode(TreeNodeF *&root, std::istream &is);
+void loadTreeNode(TreeNodeD *&root, std::istream &is);
+```
+It is also worthy to note that for algos like GLM, where the model consists of an array of weights, such a custom load/store methods are not explicitly needed.
+
+### `libcuml.so` (aka C-API wrapper over C++ API)
+As long as the guidelines in the above sub-sections have been followed, it must be easy to C-wrap the C++ API. Refer to dbscan as an example on how to properly wrap the C++ API with a C-binding. In short:
```

Comment:
I would suggest to add a link to `cpp/src/dbscan/dbscan_api.h` and `cpp/src/dbscan/dbscan_api.cpp` here.

Reply:
Done.